### ISSUE TITLE
Inverse cursor color logic is (sometimes) broken

### ISSUE THREAD

**Issue by alabuzhev** [microsoft/terminal#3647]
# Environment

```none
Windows build number:
Microsoft Windows [Version 10.0.18362.449]

Windows Terminal version (if applicable):
No.

Any other software?
No.
```

# Steps to reproduce
1. Open cmd.exe.
1. Go to window properties, "Colors" page.
2. Set "Screen Background" value to exactly `128 128 128`.

# Expected behavior
Cursor is visible no matter what background color is.

# Actual behavior
Now you see it, now you don't.

# 
Current "inverse" logic is too straightforward - it literally takes the inverse of the background, and for `128 128 128` the inverse is... also `128 128 128`. Oops.

Note: **Legacy console handles this corner case properly, as well as all Windows versions prior to 10**.

Also, legacy console inversion algorithm is not exactly "inverse" - the colors are slightly different (and subjectively more pleasant). Is it possible to get this classic inversion in the modern console please?

This issue affects a text editor app that uses console color `8` (`128 128 128` in classic color scheme) to highlight various parts of text (e.g. current line). Sudden cursor disappearing is confusing and annoying.

Example:

![image](https://user-images.githubusercontent.com/11453922/69289150-871cee00-0bf3-11ea-8893-219d107a5ca5.png)

**Comment by alabuzhev** [comment]
I've experimented with GDI renderer a bit and managed to get exactly the same cursor colors as in the Classic Console:

```DIFF
diff --git a/src/renderer/gdi/paint.cpp b/src/renderer/gdi/paint.cpp
index 2f067fc5a..2443429e4 100644
--- a/src/renderer/gdi/paint.cpp
+++ b/src/renderer/gdi/paint.cpp
@@ -733,7 +733,10 @@ bool GdiEngine::FontHasWesternScript(HDC hdc)
 
         for (auto r : cursorInvertRects)
         {
-            RETURN_HR_IF(E_FAIL, !(InvertRect(_hdcMemoryContext, &r)));
+            const auto PrevObject = SelectObject(_hdcMemoryContext, GetStockObject(LTGRAY_BRUSH));
+            const auto Result = PatBlt(_hdcMemoryContext, r.left, r.top, r.right - r.left, r.bottom - r.top, PATINVERT);
+            SelectObject(_hdcMemoryContext, PrevObject);
+            RETURN_HR_IF(E_FAIL, !Result);
         }
     }
 
```
#### Before:
![image](https://user-images.githubusercontent.com/11453922/179037063-ad319709-3c9a-489a-a3d7-0dc57ce7163a.png)
^^^ - this!

#### After:
![image](https://user-images.githubusercontent.com/11453922/179037092-1aa500c9-7a74-413f-9093-6f89dc407fc4.png)

I'm not familiar enough with other renderers, but the logic is pretty simple and equivalent to making the inverted color 25% darker:
```C++
uint32_t Color = current_cell_color();
uint32_t InvertedColor = ~Color & 0x00FFFFFF;
uint32_t CursorColor = InvertedColor ^ 0x003F3F3F;
```

**Comment by alabuzhev** [comment]
I've added DX and Atlas renderer changes to the previous comment.
Could you have a look please?
Would a PR be welcomed?

**Comment by zadjii-msft** [comment]
Lemme /cc @lhecker cause I think he's got some thoughts on the matter.

**Comment by lhecker** [comment]
Correct me if I'm wrong but the XOR results in this:
![image](https://user-images.githubusercontent.com/2256941/181643056-95a6a2e1-2f52-4d71-bd62-972c21269958.png)

But the solution for AtlasEngine is unlike that since it always subtracts 0.25 brightness so you get an intersection between input and output for `RGB(96, 96, 96)`.
So in case of AtlasEngine we could get a bit more creative... For instance this:
```hlsl
bool odd = (cellPos.x ^ cellPos.y) & 1;
float inv = odd ? 0.8 : 1.0;
color = float4(saturate(inv - color.rgb), 1);
```

This, with the help of `odd`, creates a checkerboard image (background 127 gray):
![image](https://user-images.githubusercontent.com/2256941/181647187-e5da22fe-197b-445f-8127-832ef0b94581.png)

Or default "black" (RGB = 12) background:
![image](https://user-images.githubusercontent.com/2256941/181647357-9e9f04f2-ef71-45d7-bc2d-70480b8edd31.png)

We can also use the alpha channel of the cursor texture for inversion:
```hlsl
color = float4(saturate(glyphs[cellPos].a - color.rgb), 1);
```

...and then draw a only a semi-transparent content in the glyph texture in `AtlasEngine::_drawCursor` which results in the same darkening effect.

In general I'm in favor of such a change, but I'm unsure what the best approach for the visuals is.

**Comment by alabuzhev** [comment]
@lhecker thanks, the XOR plot is correct, or, alternatively:

![image](https://user-images.githubusercontent.com/11453922/181658072-ba41650e-cf3a-49b1-ad61-9df0ecf597bf.png)

> the solution for AtlasEngine is unlike that

Mea culpa 🤦
It was supposed to be the same, but for some reason I decided to "simplify" it...
Something like this should work (and can probably be improved further):
```DIFF
diff --git a/src/renderer/atlas/shader_ps.hlsl b/src/renderer/atlas/shader_ps.hlsl
index 789cd98b0..320f2895e 100644
--- a/src/renderer/atlas/shader_ps.hlsl
+++ b/src/renderer/atlas/shader_ps.hlsl
@@ -168,6 +168,8 @@ float4 main(float4 pos: SV_Position): SV_Target
         [flatten] if (cursorColor == INVALID_COLOR && glyphs[cellPos].a != 0)
         {
             color = float4(1 - color.rgb, 1);
+            float3 shifted = ((uint3(color.rgb * 255) ^ 0x3f) & 0xff) / 255.0;
+            color = float4(shifted.rgb, 1);
         }
     }
```

> I'm unsure what the best approach for the visuals is.

I'm all for creativity and fancy visuals, but it would be nice to at least have a uniform, always readable and already tested solution everywhere first, even if not the most creative one.

**Comment by zadjii-msft** [comment]
If you'd like to submit a PR for this - go for it!

**Comment by lhecker** [comment]
BTW I'm intending to regress this issue a little bit in an upcoming version of AtlasEngine. The new rendering approach makes implementing inverted cursors, in particular the double-underline and empty-box variants, a little bit more difficult and I'd like to skip that at least in the beginning. This is because the new approach draws text in a layered approach with multiple passes and so it doesn't have the color information of each pixel readily available in order to XOR it.

So the new approach I've chosen only inspects the background color of a cell. If all components of the background color are within ±32 of #808080 gray, it chooses to consistently darken the range by -32, including the text content. In other words, it looks something like this:
![image](https://user-images.githubusercontent.com/2256941/222749124-ba07cf4f-52fd-419d-b516-8db64ad21338.png)
I could also extend the range to ±64 and darken by -64, but I felt like this already resulted in a sufficient contrast. I hope that this will be fine, if at least until I get to implement proper inversion the way we have it now.

**Comment by alabuzhev** [comment]
> within ±32 of #808080 gray

Isn't it ±16 in the picture?

**Comment by lhecker** [comment]
Ah woops. Yes it's ±16. It results in a -32 darkening, because when you invert a +16 color you get a -16 and so the delta is 32. I've considered using ±32 with a -64 darkening as well, but wasn't sure what to best pick. I was thinking that a smaller ±N could potentially result in a better visual "consistency".

**Comment by alabuzhev** [comment]
> `printf '\x1b[48;2;127;127;111m'` produces a color that makes the cursor invisible with this approach

Isn't that just a plain inversion on your screenshot? ~(127, 127, 111) = (128, 128, 144)
I'd expect a color with -32 darkening applied to be, well, darker by (-32, -32, -32), not lighter by (+1, +1, +33).

**Comment by alabuzhev** [comment]
Ah, "If **all** components of the background color are within...", and this one doesn't match.
±32 might help indeed.

**Comment by lhecker** [comment]
Yeah it's a plain inversion, because one of the components isn’t within the "±16" range anymore - at least according to the way I built it (which uses 0x80/128 as the gray value to make some bit twiddling simpler). If it was 112 instead of 111 the darkening effect would apply and the cursor would be visible. Sorry, I should’ve been more specific in my edit.
But this has shown me that my approach clearly isn’t perfect yet. I‘m still hopeful that it works if I increase it to "±32", because an inversion of `127;127;95` to `128;128;169` should hopefully be visible at least. 😅

**Comment by alabuzhev** [comment]
Alternatively, instead of "components are within the ±16 range" it could be "the distance from the center of the RGB cube is within the N range".
A sphere might capture gray-ish pixels better than a cube (with more prominent vertices and edges) and allow to specify a larger range if needed.

![image](https://user-images.githubusercontent.com/11453922/223003331-64b78883-5973-453f-b084-77abc8833cc5.png)

**Comment by lhecker** [comment]
I just realized that I can also just draw the background XOR'd with 63 (`0x3f3f3f`) where the cursor is and then draw text on top in its regular color. If I then invert the area where the cursor is (via `D3D11_BLEND_OP_SUBTRACT`), it'll result in a background color is if it was XOR'd with 192 (`0xc0c0c0`). I'm not sure if it'll be a practical problem that the text color isn't XOR'd with 63/192, but it does at least solve the issue where the cursor is invisible on top of a gray background:

![image](https://user-images.githubusercontent.com/2256941/223143346-968035ec-af24-4d15-a3f8-c0d7b9770b06.png)

Here's the current AtlasEngine for reference:

![image](https://user-images.githubusercontent.com/2256941/223144519-929b3934-f929-4426-8ba4-fe744d37157a.png)

The benefit of the current approach is that colors that are close together, stay close together. However I'm not 100% certain whether that's a good thing or not...
